<a href="https://colab.research.google.com/github/07Akshaya/Statistical-Learning-e22019/blob/main/Assignment_5_e22019.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Part A: Theoretical Fundamentals (Maximum Likelihood & Decision Space)
1. Derivation of the Bias in the MLE Covariance Matrix

The Maximum Likelihood Estimator (MLE) of the covariance matrix is


$$\widehat{\boldsymbol{\Sigma}}_{\text{MLE}} = \frac{1}{n} \sum_{i=1}^n (\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)^T$$


$\widehat{\boldsymbol{\mu}}_n = \frac{1}{n}\sum_{i=1}^n \mathbf{x}_i$



Expand the Centered Observations

$$\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n = (\mathbf{x}_i - \boldsymbol{\mu}) - (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})$$




$$\sum_{i=1}^n (\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)^T = \sum_{i=1}^n \left[ (\mathbf{x}_i - \boldsymbol{\mu}) - (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \right] \left[ (\mathbf{x}_i - \boldsymbol{\mu}) - (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \right]^T$$



$$\sum_{i=1}^n (\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)^T = \sum_{i=1}^n (\mathbf{x}_i - \boldsymbol{\mu})(\mathbf{x}_i - \boldsymbol{\mu})^T - \sum_{i=1}^n (\mathbf{x}_i - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T - \sum_{i=1}^n (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\mathbf{x}_i - \boldsymbol{\mu})^T + \sum_{i=1}^n (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T$$


$$\sum_{i=1}^n (\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)^T = \sum_{i=1}^n (\mathbf{x}_i - \boldsymbol{\mu})(\mathbf{x}_i - \boldsymbol{\mu})^T - n(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T - n(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T + n(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T$$

$$\sum_{i=1}^n (\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)^T = \sum_{i=1}^n (\mathbf{x}_i - \boldsymbol{\mu})(\mathbf{x}_i - \boldsymbol{\mu})^T - n(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T$$

Take Expectations

Applying the expectation operator,


$$\mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] = \frac{1}{n} \sum_{i=1}^n \mathbb{E}[(\mathbf{x}_i - \boldsymbol{\mu})(\mathbf{x}_i - \boldsymbol{\mu})^T] - \mathbb{E}[(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T]$$


* $\mathbb{E}[(\mathbf{x}_i - \boldsymbol{\mu})(\mathbf{x}_i - \boldsymbol{\mu})^T] = \boldsymbol{\Sigma}$
* $\mathbb{E}[(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T] = \text{Var}(\widehat{\boldsymbol{\mu}}_n) = \frac{1}{n}\boldsymbol{\Sigma}$



$$\mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] = \frac{1}{n} (n\boldsymbol{\Sigma}) - \frac{1}{n}\boldsymbol{\Sigma} = \boldsymbol{\Sigma} - \frac{1}{n}\boldsymbol{\Sigma} = \left(\frac{n-1}{n}\right)\boldsymbol{\Sigma}$$

which proves that the MLE covariance estimator is biased downward.

**Bessel's Correction Elimination of Bias**

To remove the bias, multiply the MLE estimator by the reciprocal of the bias factor


$$\mathbf{S} = \frac{n}{n-1} \widehat{\boldsymbol{\Sigma}}_{\text{MLE}} = \frac{n}{n-1} \left( \frac{1}{n} \sum_{i=1}^n (\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)^T \right) = \frac{1}{n-1} \sum_{i=1}^n (\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)^T$$

Evaluating the expectation confirms its lack of bias:


$$\mathbb{E}[\mathbf{S}] = \frac{n}{n-1} \mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] = \frac{n}{n-1} \left(\frac{n-1}{n}\right)\boldsymbol{\Sigma} = \boldsymbol{\Sigma}$$

Operational Error Diagnostics

Suppose

$H_0:\ \text{Structure is healthy}$

and

$H_1:\ \text{Structure is damaged or degraded}.$



### Type I Error ($\alpha$)

A Type I Error occurs when $(H_0)$ is true but is rejected.

In structural health monitoring:

$\boxed{ \alpha = P(\text{Declare damage}\mid \text{Structure healthy}) }$

This corresponds to a **false alarm**, causing unnecessary inspections, shutdowns, or maintenance actions.



### Type II Error ($\beta$)

A Type II Error occurs when $(H_1)$ is true but $(H_0)$ is not rejected.

In structural health monitoring:

$\boxed{ \beta = P(\text{Declare healthy}\mid \text{Structure damaged}) }$

This corresponds to a **missed detection**, potentially allowing crack growth or structural failure to go unnoticed.


### Statistical Power

The power of a diagnostic test is

$\boxed{ \text{Power}=1-\beta }$

which is the probability of correctly detecting damage when damage actually exists.



### Effect of an Ultra-Conservative Threshold ($\alpha=0.0001$)

Reducing $(\alpha)$ means that the rejection region is moved further into the tails of the reference distribution.

For example, in a standard normal test:

* $(\alpha=0.05)$ gives a critical value near $(z=1.96)$
* $(\alpha=0.0001)$ gives a critical value near $(z=3.89)$

Thus much larger deviations are required before an alarm is triggered.



### Consequence for Power

Because the rejection region becomes smaller,

$\beta \uparrow$

and therefore

$1-\beta \downarrow.$

So:

$\boxed{ \alpha \downarrow \quad\Longrightarrow\quad \beta \uparrow \quad\Longrightarrow\quad \text{Power} \downarrow }$

The monitoring system becomes less sensitive to small or emerging structural defects.



### Geometric Interpretation: Confidence Ellipsoid

For multivariate monitoring, healthy operation is often represented by a confidence ellipsoid:

${x:(x-\mu)^T\Sigma^{-1}(x-\mu)\le c_\alpha}.$

$(x-\mu)^T\Sigma^{-1}(x-\mu)=c$

The threshold $(c_\alpha)$ is determined from a chi-square distribution:

$c_\alpha = \chi^2_{m,,1-\alpha}.$

As $(\alpha)$ decreases,

$1-\alpha$

increases, and therefore the chi-square quantile increases:

$c_\alpha \uparrow.$

Since the ellipsoid volume satisfies

$V \propto c_\alpha^{m/2}\sqrt{\det(\Sigma)},$

a larger $(c_\alpha)$ produces a larger volume.

Thus

$\boxed{ \alpha\downarrow \Longrightarrow c_\alpha\uparrow \Longrightarrow V_{\text{ellipsoid}}\uparrow }$

The healthy-region boundary expands and encloses more observations.



### Engineering Interpretation

An ultra-conservative threshold such as

$\alpha = 0.0001$

greatly reduces false shutdowns and nuisance alarms. However:

* False alarms (Type I errors) become extremely rare.
* Missed detections (Type II errors) increase.
* Statistical power decreases.
* The healthy-operation confidence ellipsoid expands.
* Small structural changes become harder to distinguish from normal variability.

Therefore, choosing $(\alpha)$ is fundamentally a trade-off between **operational reliability** and **damage-detection sensitivity**.

### Slutsky’s Theorem (Multivariate Form)

Let $(X_n)$ be a sequence of random vectors and $(Y_n)$ a sequence of random matrices. If

$X_n \xrightarrow{d} X$

and

$Y_n \xrightarrow{p} C,$

where $(C)$ is a constant matrix, then

$Y_nX_n \xrightarrow{d} CX.$

Similarly,

$X_n+Y_n \xrightarrow{d} X+C.$

If $(C)$ is nonsingular, then

$Y_n^{-1}X_n \xrightarrow{d} C^{-1}X.$

---

### Justification for Replacing $(\Sigma)$ by $(S)$

The Lindeberg–Lévy Multivariate Central Limit Theorem states that

$\sqrt{n}(\hat{\mu}_n-\mu) \xrightarrow{d} N(0,\Sigma).$

Equivalently,

$\Sigma^{-1/2}\sqrt{n}(\hat{\mu}_n-\mu) \xrightarrow{d} N(0,I).$

The sample covariance matrix

$S=\frac{1}{n-1}\sum_{i=1}^{n}(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T$

is a consistent estimator of $(\Sigma)$. By the Weak Law of Large Numbers,

$S \xrightarrow{p} \Sigma.$

Since the matrix inverse square-root is a continuous function on positive-definite matrices,

$S^{-1/2}\xrightarrow{p}\Sigma^{-1/2}.$

Applying Slutsky’s Theorem to

$\sqrt{n}(\hat{\mu}_n-\mu)\xrightarrow{d}N(0,\Sigma)$

and

$S^{-1/2}\xrightarrow{p}\Sigma^{-1/2},$

gives

$S^{-1/2}\sqrt{n}(\hat{\mu}_n-\mu) \xrightarrow{d} \Sigma^{-1/2}N(0,\Sigma).$

The covariance of the limiting distribution is

$\Sigma^{-1/2}\Sigma(\Sigma^{-1/2})^T=I,$

hence

$S^{-1/2}\sqrt{n}(\hat{\mu}_n-\mu) \xrightarrow{d} N(0,I).$

Therefore, replacing the unknown population covariance matrix $(\Sigma)$ by the consistent estimator $(S)$ does not change the asymptotic distribution. Consequently,

$\sqrt{n}(\hat{\mu}_n-\mu) \overset{a}{\sim} N(0,S),$

which implies

$\boxed{\hat{\mu}_n \overset{a}{\sim} N\left(\mu,\frac{1}{n}S\right)}.$

Thus, Slutsky’s Theorem rigorously justifies the practical substitution of $(S)$ for the latent covariance matrix $(\Sigma)$ in asymptotic parametric modelling.

To verify the first-moment homogeneity of the industrial dataset, the following Python function implements the Multivariate Analysis of Variance (MANOVA) framework using the matrix decompositions and asymptotic transformations detailed in the sources.

In [3]:
### **Python Implementation**
import numpy as np
import pandas as pd
import scipy.stats as stats

def verify_first_moment_homogeneity(df: pd.DataFrame, g_chunks: int = 5) -> dict:
    """
    Partitions the dataset into g chunks and evaluates first-moment homogeneity
    via Wilks' Lambda and Bartlett's Chi-Square asymptotic transformation.
    """
    n, m = df.shape
    nj = n // g_chunks  # Size of each consecutive block (e.g., 1,000)

    # 1. Split df into g equal consecutive blocks
    chunks = [df.iloc[j*nj : (j+1)*nj].values for j in range(g_chunks)]

    # 2. Compute the global mean (mu_n) and individual chunk means (mu_j)
    global_mean = df.mean().values
    chunk_means = [np.mean(chunk, axis=0) for chunk in chunks]

    # 3. Calculate Within-Chunk Matrix (W) and Between-Chunk Matrix (B)
    W = np.zeros((m, m))
    B = np.zeros((m, m))

    for j in range(g_chunks):
        # Within-chunk variation: measures random variation around chunk centers
        diff_w = chunks[j] - chunk_means[j]
        W += diff_w.T @ diff_w

        # Between-chunk variation: measures displacement of chunk centers (drift)
        diff_b = (chunk_means[j] - global_mean).reshape(-1, 1)
        B += nj * (diff_b @ diff_b.T)

    # 4. Compute Wilks' Lambda, Bartlett's Chi-Square, and the final p-value
    det_W = np.linalg.det(W)
    det_total = np.linalg.det(W + B)
    wilks_lambda = det_W / det_total  # Lambda = |W| / |B + W|

    # Bartlett transformation:
    # chi2_calc = -(n - 1 - (m + g)/2) * ln(Lambda)
    chi2_calc = -(n - 1 - (m + g_chunks) / 2.0) * np.log(wilks_lambda)

    # Degrees of freedom: m * (g - 1)
    df_chi2 = m * (g_chunks - 1)
    p_value = 1 - stats.chi2.cdf(chi2_calc, df_chi2)

    return {
        "Wilks_Lambda": wilks_lambda,
        "Bartlett_Chi2": chi2_calc,
        "p_value": p_value,
        "df": df_chi2
    }

# Execution
results = verify_first_moment_homogeneity(df_strain, g_chunks=5)
print(results)

{'Wilks_Lambda': np.float64(0.9904519796929591), 'Bartlett_Chi2': np.float64(47.9215049961488), 'p_value': np.float64(3.2255259508895406e-06), 'df': 12}



### **Numerical Results**
Based on the execution of the code with the provided synthetic data:
*   **Wilks' Lambda ($\Lambda$):** ~0.99045
*   **Bartlett $\chi^2$ Statistic:** ~47.9215
*   **Degrees of Freedom ($df$):** 12 ($3 \text{ features} \times (5 \text{ chunks} - 1)$)
*   **p-value:** ~3.23e-06

### **Diagnostic Conclusion**
Given an alpha threshold of **$\alpha = 0.05$**, the calculated **p-value (~0.0000032)** is significantly lower than the threshold. According to the decision rule, we **reject the null hypothesis ($H_0$)** that the chunk population means are equivalent.

**Conclusion:** The **baseline center has shifted over time**, and the first moment is **not homogeneous**. This result successfully identifies the "subtle, hidden structural drift" injected into the final 1,000 snapshots of the synthetic data (where Sensor 1 and Sensor 3 values were modified). Therefore, the weak i.i.d. assumption is violated, and further data stabilization or drift correction is required before initializing global PCA or Factor Analysis layers.

### **Part A: Theoretical Fundamentals (Coordinate Projections & Orthogonality)**

#### **1. Proof of Covariance Diagonalization**
To prove that the theoretical covariance matrix of the transformed vector $Z_i$ simplifies to $\Lambda$, we follow these linear algebra steps:

*   **Definition**: We start with the orthogonal projection mapping $Z_i = P^T \tilde{X}_i$.
*   **Expansion**: The covariance matrix is defined as $E[Z_i Z_i^T]$. Substituting the mapping, we get:
    $$E[Z_i Z_i^T] = E[(P^T \tilde{X}_i)(P^T \tilde{X}_i)^T]$$.
*   **Transpose Identity**: Utilizing the property $(AB)^T = B^T A^T$, the expression becomes:
    $$E[P^T \tilde{X}_i \tilde{X}_i^T P]$$.
*   **Linearity of Expectation**: Since $P$ is a constant matrix, we pull it out of the expectation:
    $$P^T E[\tilde{X}_i \tilde{X}_i^T] P = P^T \Sigma P$$..
*   **Spectral Substitution**: Substituting the spectral decomposition $\Sigma = P \Lambda P^T$ into the equation:
    $$P^T (P \Lambda P^T) P$$..
*   **Orthogonality Simplification**: Because $P$ is an orthogonal matrix, $P^T P = P P^T = I$. Thus:
    $$(P^T P) \Lambda (P^T P) = I \Lambda I = \Lambda$$..

**Geometric and Statistical Meaning**: The diagonal nature of $\Lambda$ means that the **cross-covariances between distinct coordinates $Z_{i,j}$ and $Z_{i,k}$ (where $j \neq k$) are exactly zero**. Statistically, this proves the transformation **completely de-correlates the feature space**. Geometrically, it signifies that the elliptical density contours of the data have been rotated from an arbitrary alignment into a **perfect alignment with the principal axes**.

#### **2. Proof of Total Variance Preservation**
Using the algebraic property that the **trace is invariant under cyclic permutations** ($tr(ABC) = tr(BCA)$), we can prove variance is preserved:

*   Total variance in raw space is $tr(\Sigma)$.
*   Substituting the spectral decomposition: $tr(\Sigma) = tr(P \Lambda P^T)$.
*   Applying cyclic permutation: $tr(P (\Lambda P^T)) = tr((\Lambda P^T) P) = tr(\Lambda (P^T P))$.
*   Since $P^T P = I$, the expression simplifies to:
    $$tr(\Lambda I) = tr(\Lambda) = \sum_{j=1}^m \lambda_j$$..

#### **3. Subspace Variance Ratios**
When partitioning the matrix into explained ($\hat{P}_k$) and residual ($\hat{P}_{m-k}$) subspaces, the ratios are defined as:
*   **Cumulative Explained Variance Ratio $\Phi(k)$**: Measures the informational content of the first $k$ components:
    $$\Phi(k) = \frac{\sum_{j=1}^k \hat{\lambda}_j}{\sum_{i=1}^m \hat{\lambda}_i}$$.
*   **Residual Unexplained Variance Ratio $\Psi(k)$**: Gauges the scale of the unmodeled operational subspace:
    $$\Psi(k) = 1.0 - \Phi(k) = \frac{\sum_{j=k+1}^m \hat{\lambda}_j}{\sum_{i=1}^m \hat{\lambda}_i}$$.

#### **4. Subspace Energy Decomposition**
*   **Reconstructed Vector ($\hat{x}_i$)**: The projection of low-dimensional scores back into the original space: $\hat{x}_i \triangleq \hat{\mu}_n + \hat{P}_k z_{i,k}$.
*   **Residual Error Vector ($e_i$)**: The difference between the raw and reconstructed vectors: $e_i = x_i - \hat{x}_i = \hat{P}_{m-k} z_{i,m-k}$.

**Pythagorean Proof**: Because $\hat{P}_k$ and $\hat{P}_{m-k}$ span **mutually orthogonal subspaces**, their inner product is zero. The total squared Euclidean norm of the centered observation vector $(x_i - \hat{\mu}_n)$ decomposes as follows:
$$\|x_i - \hat{\mu}_n\|^2 = \|\hat{P}_k z_{i,k} + \hat{P}_{m-k} z_{i,m-k}\|^2$$..
Due to orthogonality, the cross-terms vanish, leaving:
$$\|x_i - \hat{\mu}_n\|^2 = \|z_{i,k}\|^2 + \|z_{i,m-k}\|^2$$..

#### **5. Diagnostic Functions and Event Matching**
*   **Hotelling's $T^2$ Statistic**: Measures Mahalanobis distance within the $k$-dimensional principal subspace. It is designed to detect shifts in **nominal, high-variance operational modes**.
*   **Residual $Q$ Statistic (SPE)**: Measures the magnitude of the residual vector. It detects a **break in the underlying structural correlation**.

**Event Matching**:
*   **Environmental Load Anomaly (e.g., wind loads)** $\rightarrow$ **Hotelling's $T^2$**: These anomalies typically represent extreme versions of existing operational modes. The system remains within its known physical correlation structure but moves to an extreme point within the principal subspace.
*   **Internal Structural Fracture (e.g., fatigue crack)** $\rightarrow$ **Residual $Q$ (SPE)**: A fracture introduces a **new, unmodeled physical phenomenon**. It changes the load pathways, causing the sensors to relate to each other in a way the baseline model cannot explain, thereby projecting energy into the **residual noise floor**.

In [ ]:
### **Python Optimization Pipeline**

import numpy as np
import pandas as pd

def analyze_pca_subspace_diagnostics(X_raw: np.ndarray, k_range) -> pd.DataFrame:
    """
    Optimizes PCA truncation boundaries by analyzing Hotelling's T2
    and Q (SPE) statistics across different subspace sizes.
    """
    n, m = X_raw.shape

    # 1. Standardize and Center the incoming realizations
    # Applying Z-score mapping using unbiased sample standard deviation (ddof=1)
    mu_hat = np.mean(X_raw, axis=0)
    sigma_hat = np.std(X_raw, axis=0, ddof=1)

    # Numerical guardrail for zero-variance channels
    sigma_hat[sigma_hat == 0] = 1e-15
    X_standardized = (X_raw - mu_hat) / sigma_hat

    # 2. Decompose the unbiased sample covariance matrix S
    S = np.cov(X_standardized, rowvar=False, ddof=1)
    lambdas_raw, P_raw = np.linalg.eigh(S)

    # Sort components in descending order (lambda_1 >= ... >= lambda_m)
    idx = np.argsort(lambdas_raw)[::-1]
    lambdas = lambdas_raw[idx]
    P = P_raw[:, idx]

    # Map realizations to Principal Component Scores: zi = P^T * (xi - mu)
    Z_scores = X_standardized @ P

    metrics = []

    # 3. Loop through subspace cutoff values k
    for k in k_range:
        # Calculate Hotelling's T2 statistic for each sample
        # Formula: T2_i = sum_{j=1}^k (z_{i,j}^2 / lambda_j)
        T2_samples = np.sum((Z_scores[:, :k]**2) / lambdas[:k], axis=1)

        # Calculate Q statistic (SPE) for each sample
        # Formula: Q_i = sum_{j=k+1}^m (z_{i,j}^2)
        Q_samples = np.sum(Z_scores[:, k:]**2, axis=1)

        # 4. Compute expected values (empirical means)
        metrics.append({
            "k_truncation": k,
            "Mean_Hotelling_T2": np.mean(T2_samples),
            "Mean_Q_Statistic": np.mean(Q_samples)
        })

    return pd.DataFrame(metrics)

# Example Execution
# results = analyze_pca_subspace_diagnostics(df_baseline.values, k_range=)

### **Theoretical Verification Breakdown**

*   **Hotelling's $T^2$ Expected Value**: In the principal subspace, the expected value of $T^2$ should theoretically approach $k$. This is because the transformation de-correlates the variables such that $E[z_j^2] = \lambda_j$, causing the term $\sum (z_j^2/\lambda_j)$ to behave as a sum of unit-variance components.
*   **$Q$ Statistic (SPE) Expected Value**: The $Q$ statistic measures the energy remaining in the **residual noise floor**. Its empirical mean should equal the sum of the discarded eigenvalues: $E[Q] = \sum_{j=k+1}^m \lambda_j$.
*   **Subspace Optimization**: By tracking the decay of $Q$ against $k$, engineers can identify the "flat elbow" where the deterministic mechanical behavior transitions into unmodeled stochastic white noise.
*   **Diagnostic Sensitivity**: **Hotelling's $T^2$** is most sensitive to **environmental load anomalies** (shifting within known modes), while the **$Q$ statistic** is the primary indicator for **internal structural fractures** that break the underlying correlation pathways.

### Part A: Theoretical Fundamentals (Generative Model & Commonalities)

#### 1. Proof of the Fundamental Equation of Factor Analysis
The Factor Analysis (FA) model defines the normalized feature vector as **$Z_i = \Lambda F_i + \epsilon_i$**. To derive the population correlation matrix **$R$**, we evaluate the expectation of the outer product of $Z_i$:
$$R = E[Z_i Z_i^T] = E[(\Lambda F_i + \epsilon_i)(\Lambda F_i + \epsilon_i)^T]$$

Expanding this product yields:
$$R = \Lambda E[F_i F_i^T] \Lambda^T + \Lambda E[F_i \epsilon_i^T] + E[\epsilon_i F_i^T] \Lambda^T + E[\epsilon_i \epsilon_i^T]$$

Applying the model's orthogonal and independence constraints:
*   **Orthogonal Factors**: $E[F_i F_i^T] = I_{k \times k}$.
*   **Subspace Orthogonality**: $E[\epsilon_i F_i^T] = 0$ (and its transpose is also 0).
*   **Diagonal Noise**: $E[\epsilon_i \epsilon_i^T] = \Psi$.

Substituting these into the expanded equation results in the **Fundamental Equation**:
**$R = \Lambda \Lambda^T + \Psi$**.

#### 2. Definition and Physical Interpretation of $h_j^2$ and $\phi_j^2$
For any $j$-th sensor (diagonal entry), the total unit variance is partitioned as:
**$1.0 = \sum_{r=1}^k \lambda_{j,r}^2 + \phi_j^2$**.

*   **Communality ($h_j^2$)**: Defined as **$\sum_{r=1}^k \lambda_{j,r}^2$**, it represents the portion of a sensor's variance explained by the $k$ shared structural factors. In diagnostics, a high $h_j^2$ indicates the sensor is effectively capturing the global physical mechanics of the asset.
*   **Uniqueness ($\phi_j^2$)**: Defined as **$1.0 - h_j^2$**, it represents variance isolated to that specific channel, including localized anomalies and electronic white noise.
*   **Diagnostic Application**:
    *   **Structural Damage**: Causes wide-scale realignments in the factor loading matrix and the communality spectrum across multiple sensors.
    *   **Sensor Malfunction**: Characterized by a **precipitous drop in communality** and a corresponding **spike in uniqueness toward 1.0** for only the faulty sensor, allowing for immediate isolation from true structural shifts.

#### 3. Principal Component Loadings vs. Varimax-Rotated Factor Loadings
*   **Raw PCA Loadings**: Follow a strict mathematical hierarchy where each component tries to sweep up as much mixed variance as possible, often resulting in complex "mathematical jumbles" where sensors load across multiple axes.
*   **Varimax-Rotated FA Loadings**: Apply an orthogonal rotation to achieve a **"simple structure."** This maximizes the variance of the loadings within columns, isolating signals so that specific sensors group cleanly into distinct, interpretable physical factors (e.g., Sensors 1 and 2 on Factor 1, and Sensor 3 on Factor 2).
*   **Invariance of Rotations**: Orthogonal rotations (using a matrix $T$) can alter individual loading columns ($\Lambda^* = \Lambda T$) without changing $h_j^2$ or the global approximation $R \approx \Lambda \Lambda^T + \Psi$. This is because $(\Lambda T)(\Lambda T)^T = \Lambda T T^T \Lambda^T = \Lambda I \Lambda^T = \Lambda \Lambda^T$, preserving the total shared variance captured by the model.

---

### Part B: Estimation of Factor Scores (Thomson’s Method)

#### 1. Joint Subspace Covariance Construction
To derive factor scores using conditional expectations, we first construct the joint partitioned vector **$Y_i = [F_i^T, Z_i^T]^T$**. The joint population covariance matrix **$\Sigma_{YY}$** is composed of four blocks:
1.  **$E[F_i F_i^T] = I_{k \times k}$**: Orthogonal common factor constraint.
2.  **$E[F_i Z_i^T] = \Lambda^T$**: The transpose of the factor loading matrix.
3.  **$E[Z_i F_i^T] = \Lambda$**: The factor loading matrix.
4.  **$E[Z_i Z_i^T] = R$**: The visible correlation matrix.

Arranging these blocks for the joint vector $Y_i$ yields:
**$\Sigma_{YY} = \begin{bmatrix} I_{k \times k} & \Lambda^T \\ \Lambda & R \end{bmatrix}$**.

*Note: If the vector is ordered as $Y_i = [Z_i^T, F_i^T]^T$, the matrix expands to $\Sigma_{YY} = \begin{bmatrix} R & \Lambda \\ \Lambda^T & I_{k \times k} \end{bmatrix}$.*

#### 2. Derivation of Thomson’s Estimator
Using the conditional expectation theorem for Multivariate Normal systems, the realization $f_i$ is found via:
**$E[X_1 | x_2] = \mu_1 + \Sigma_{12} \Sigma_{22}^{-1} (x_2 - \mu_2)$**.

Mapping the FA parameters to this identity (where $X_1 = F_i$ and $X_2 = Z_i$):
*   Means $\mu_1$ and $\mu_2$ are **0**.
*   The cross-covariance $\Sigma_{12}$ is **$\Lambda^T$**.
*   The observed covariance $\Sigma_{22}$ is **$R$**.

Substituting these into the projection formula yields the final score matrix equation for **Thomson’s estimator**:
**$f_i = \Lambda^T R^{-1} z_i$**.

This estimator is the **Best Linear Unbiased Predictor (BLUP)**, as it minimizes the mean squared error between the hidden physical state and the feature estimates.

To implement the **Parametric Factor Partition Engine**, the diagnostic pipeline is structured into isolated algorithmic layers designed for robust structural health monitoring and sensor error isolation.

### 1. Z-Score Mapping Layer
The engine first isolates numeric observation columns from an arbitrary data frame and maps them into a standard normal space **$Z \in \mathbb{R}^{n \times m}$**. The standardization process uses the **unbiased sample standard deviation** ($ddof=1$) to ensure statistical accuracy in finite samples.

To maintain runtime stability, an explicit **numerical guardrail** is implemented:
*   The algorithm identifies zero-variance channels ($\sigma_j = 0$) where a sensor might be stuck or providing a flat signal.
*   These channels are capped at a minimum boundary threshold of **$10^{-15}$**. This prevents divide-by-zero exceptions during the Z-score calculation: $Z = \frac{X - \hat{\mu}}{\hat{\sigma}}$.

### 2. Dimensionality Constraint Layer
Before decomposition, the engine enforces two strict validation rules to ensure the mathematical validity of the Factor Analysis model:
*   **Snapshot Count Check**: The system rejects processing if the number of operational snapshots ($n$) is less than or equal to the sensor channel count ($m$). This ensures the sample size is sufficient to isolate principal or latent components.
*   **Latent Space Boundary**: The requested latent space dimension ($k$) must be **strictly less than** the available feature space dimension ($m$). This constraint is necessary because Factor Analysis seeks to explain the $m \times m$ correlation matrix using a compressed $k$-dimensional structural subspace.

### 3. Information Decomposition Layer
The core engine fits a Factor Analysis framework utilizing either Maximum Likelihood or SVD algorithms, integrated with an **orthogonal Varimax rotation**. This rotation is critical for interpretability, as it maximizes the loading variance within columns to create a "simple structure" where sensors group into distinct physical categories.

From the fitted model, the engine computes two primary vector arrays used for diagnostics:
*   **Communality Vector ($h^2 \in \mathbb{R}^m$)**: This represents the total shared variance captured across each channel. Mathematically, it is the sum of the squared loadings for each sensor across all $k$ factors: $h_j^2 = \sum_{r=1}^k \lambda_{j,r}^2$.
*   **Uniqueness Vector ($\phi^2 \in \mathbb{R}^m$)**: This represents the remaining isolated sensor noise floor ($1.0 - h^2$).

### 4. Diagnostic Application
In an asset diagnostics system, these partitioned vectors allow for the immediate isolation of malfunctions. A **structural failure** (such as a mechanical load shift) will cause wide-scale realignments in the communality spectrum across multiple sensors. Conversely, a **sensor-specific failure** or decalibration is identified when a single channel's communality ($h^2$) drops precipitously while its uniqueness ($\phi^2$) spikes toward 1.0, indicating the signal has decoupled from the shared physical state of the asset.

To implement the **Latent Score Projection** and **Performance Diagnostic Telemetry** layers of the Factor Analysis pipeline, you must utilize the following mathematical and algorithmic structures derived from the sources:

### 1. Mathematical State Reconstruction (Thomson’s MMSE)
The transformation of the standard normal snapshot array ($Z \in \mathbb{R}^{n \times m}$) into the hidden factor states ($F \in \mathbb{R}^{n \times k}$) is achieved via **Thomson’s Regression Method**. This method treats the factor scores as the **Minimum Mean Squared Error (MMSE)** estimator, derived from the conditional expectation $E[F_i | z_i]$.

Using the system relation provided in the requirements, the scores are projected as:
$$F = Z R^{-1} \Lambda_{rotated}$$

*   **$Z$**: The $n \times m$ matrix of standardized (Z-score mapped) observations.
*   **$R^{-1}$**: The inverse of the empirical correlation matrix.
*   **$\Lambda_{rotated}$**: The factor loading matrix after applying an orthogonal Varimax rotation to achieve "simple structure".

This specific linear transformation is the **Best Linear Unbiased Predictor (BLUP)**, which minimizes the mean squared error between the estimated states and the true unobservable physical phenomena.

### 2. Performance Diagnostic Telemetry
Once the factor scores $F$ are reconstructed, the engine evaluates the distribution of energy across the latent subspace to verify model performance.

*   **Empirical Variance Calculation**: The variance for each of the $k$ factors is calculated using an **unbiased scale correction** ($ddof=1$) to account for finite sample snapshots:
    $$\text{Var}(F_{r}) = \frac{1}{n-1} \sum_{i=1}^{n} (f_{i,r} - \bar{f}_r)^2$$
*   **System-Wide Metrics Reporting**: To confirm the health of the diagnostic layer, the engine prints a breakdown of the captured structural energy versus the localized noise floor. The calculations are as follows:

| Metric | Mathematical Formula | Physical Meaning |
| :--- | :--- | :--- |
| **Average System Communality %** | $\frac{1}{m} \sum_{j=1}^m h_j^2 \times 100$ | The average percentage of sensor variance explained by shared structural dynamics. |
| **Average System Uniqueness %** | $\frac{1}{m} \sum_{j=1}^m \phi_j^2 \times 100$ | The average localized noise floor or specific sensor error across the entire array. |

### 3. Diagnostic Interpretation
In a production environment, these metrics provide immediate insight into asset health:
*   **Structural Integrity**: High average communality confirms that the sensors are well-coupled to the underlying physical modes.
*   **Sensor Reliability**: A sudden increase in average uniqueness, particularly if driven by a single sensor's $h_j^2$ dropping precipitously, allows the system to **isolate instrument malfunctions** (like electrical shorts) from genuine structural damage.

The **2 × 2 Factor Diagnostics Grid Matrix** is engineered for production-grade structural health monitoring, utilizing a unified **Plotly Subplots** architecture that enforces specific geometric constraints for maximum legibility and analytical precision.

### **Global Layout and Grid Specifications**
*   **Canvas Geometry:** The dashboard is rendered on a global canvas with a **width of 1250** and **height of 750**, applying the **`plotly_white`** theme.
*   **Spacing and Padding:** To prevent overlapping labels and visual clutter, the matrix is configured with an explicit **`horizontal_spacing=0.24`** and **`vertical_spacing=0.28`**.
*   **Marginal Insulation:** Clean margin structures isolate the interface elements: **top (150)**, **bottom (60)**, **left (140)**, and **right (80)**.
*   **Legend Architecture:** A single, **horizontally oriented legend** is centered directly above the subplots (`orientation="h", x=0.5, y=1.02`).

### **The Four Diagnostic Tracking Profiles**

#### **1. Structural Loadings Matrix Heatmap [Row 1, Col 1]**
This subplot renders the absolute values of the rotated component weights matrix (**$|\Lambda_{rotated}|$**) to reveal sensor sensitivity to latent physical modes.
*   **Color Scale:** Uses a continuous **`YlOrRd`** (Yellow-Orange-Red) palette.
*   **Colorbar Isolation:** To protect the clarity of y-axis sensor identifiers, the scale bar is isolated to the left side: `x=-0.15, len=0.38, y=0.78, yanchor="middle", xanchor="right"`.

#### **2. Variance Partitioning Profile [Row 1, Col 2]**
A **horizontal stacked bar chart** maps the variance allocation for each sensor channel, allowing for immediate identification of signal quality.
*   **Shared Communality ($h^2 \times 100$):** The base layer represents structural energy using color hex code **`#1f77b4`** (blue).
*   **Isolated Uniqueness ($\phi^2 \times 100$):** This layer represents the noise floor, stacked directly on top using color hex code **`#ff7f0e`** (orange).
*   **Boundary Constraints:** The x-axis is strictly limited between **** to represent total percentage space.

#### **3. Sensor Uniqueness Line Profile [Row 2, Col 1]**
A clear line chart tracks the isolated channel noise thresholds (**$\phi^2$**) across monitored hardware positions to isolate specific instrument faults.
*   **Visual Parameters:** Constructed with a **red (`#d62728`)**, 2-pixel wide **dot-dashed line** and visual **"x" indicators**.
*   **Label Optimization:** X-axis category ticks are rotated to a **`tickangle=25`** to ensure readability across high-density sensor arrays.

#### **4. Latent Factor Variance Distribution [Row 2, Col 2]**
A vertical bar chart verifies the post-rotation energy level allocations for the extracted factor score vectors.
*   **Visual Styling:** Bars are rendered in **green (`#2ca02c`)** with an explicit **black boundary** envelope of **0.5 width**.
*   **Diagnostic Purpose:** This verifies that the extracted factors maintain significant empirical variance after the Varimax rotation.

In [4]:
import numpy as np
import pandas as pd
from typing import Optional, Sequence, Dict, Any
from sklearn.decomposition import FactorAnalysis
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# =====================================================================
# COMBINED SUBSPACE ENGINES (PCA & FA PIPELINES)
# =====================================================================

class DualSubspaceDiagnosticsEngine:
    def __init__(self, df: pd.DataFrame):
        self.df = df

    def compute_empirical_pca(self, columns: Optional[Sequence[str]] = None, show_plot: bool = True) -> Dict[str, Any]:
        """
        Operationalizes the geometric de-correlation framework via PCA.
        Decomposes the unbiased empirical covariance matrix S into its orthogonal
        basis P_hat and renders a clean 2x3 optimization and feature contribution dashboard.
        """
        if self.df is None:
            raise ValueError("Error: No data loaded.")

        if columns is None:
            target_cols = self.df.select_dtypes(include=[np.number]).columns.tolist()
            if 'count' in target_cols: target_cols.remove('count')
            target_cols = list(target_cols)
        else:
            target_cols = list(columns)

        X = self.df[target_cols].copy().dropna().values
        n, m = X.shape

        if n <= m:
            raise ValueError("Sample size n must be larger than feature dimensions m to isolate principal components.")

        # 1. Centering (Crucial for proper Covariance decomposition)
        mu_hat = np.mean(X, axis=0)
        X_centered = X - mu_hat

        # 2. Compute Unbiased Sample Covariance Matrix S (ddof=1)
        S_matrix = np.cov(X, rowvar=False, ddof=1)

        # 3. Spectral Decomposition
        eigenvalues, eigenvectors = np.linalg.eigh(S_matrix)
        idx = np.argsort(eigenvalues)[::-1]
        lambda_hat = eigenvalues[idx]
        P_hat = eigenvectors[:, idx]
        lambda_hat = np.clip(lambda_hat, a_min=1e-15, a_max=None)

        # 4. Variance Information Calculations
        total_variance = np.sum(lambda_hat)
        if total_variance == 0:
            raise ValueError("Total variance is zero.")

        explained_variance_ratio = lambda_hat / total_variance
        cumulative_variance_ratio = np.cumsum(explained_variance_ratio)
        unexplained_variance_ratio = 1.0 - cumulative_variance_ratio

        # 5. Map Realizations to Principal Component Scores
        Z_scores = np.dot(X_centered, P_hat)

        # 6. Evaluate Statistical Distance Metrics (T^2 and Q) across truncation bounds
        k_range = np.arange(1, m + 1)
        mean_T2_vs_k = []
        mean_Q_vs_k = []
        T2_matrix = np.zeros((n, len(k_range)))
        Q_matrix = np.zeros((n, len(k_range)))

        for idx_k, k in enumerate(k_range):
            # Hotelling's T^2 calculation for k components
            Z_k = Z_scores[:, :k]
            lambda_k = lambda_hat[:k]
            T2_samples = np.sum((Z_k ** 2) / lambda_k, axis=1)
            T2_matrix[:, idx_k] = T2_samples
            mean_T2_vs_k.append(np.mean(T2_samples))

            # Q-Statistic (SPE) calculation for truncation cutoff k
            if k < m:
                Z_residual = Z_scores[:, k:]
                Q_samples = np.sum(Z_residual ** 2, axis=1)
            else:
                Q_samples = np.zeros(n)
            Q_matrix[:, idx_k] = Q_samples
            mean_Q_vs_k.append(np.mean(Q_samples))

        print(f"\n--- Operationalizing Geometric De-correlation Layer (PCA) ---")
        print(f"Decomposing structural space of {m} features using {n} samples.")
        print(f"Total System Variance (Trace[S]): {total_variance:.4f}")

        # 7. Enhanced 2x3 Diagnostic Dashboard
        if show_plot:
            pc_labels = [f"PC {i+1}" for i in range(m)]
            k_labels = [f"k={k}" for k in k_range]
            sensor_labels = target_cols

            fig = make_subplots(
                rows=2, cols=3,
                horizontal_spacing=0.12,
                vertical_spacing=0.25,
                subplot_titles=(
                    "Feature Loading Matrix |P_hat|",
                    "Component Values (Eigenvalues λ)",
                    "Information Profile (Explained Var.)",
                    "Residual Space (Unexplained Var.)",
                    "Mean Hotelling's T² vs Subspace Size k",
                    "Mean Q Statistic (SPE) vs Cutoff k"
                )
            )

            # --- ROW 1, COL 1: PCA Loading Heatmap ---
            fig.add_trace(
                go.Heatmap(
                    z=np.abs(P_hat),
                    x=pc_labels,
                    y=sensor_labels,
                    colorscale='YlOrRd',
                    colorbar=dict(
                        title="Weight",
                        x=0.24,
                        len=0.4,
                        y=0.8,
                        yanchor="middle",
                        xanchor="right"
                    ),
                    showscale=True
                ),
                row=1, col=1
            )
            fig.update_xaxes(title_text="Principal Axes", row=1, col=1)

            # --- ROW 1, COL 2: Absolute Eigenvalues ---
            fig.add_trace(go.Bar(x=pc_labels, y=lambda_hat, name="Eigenvalue (λ_j)", marker=dict(color='#1f77b4'), showlegend=False), row=1, col=2)
            fig.update_yaxes(title_text="Variance Magnitude", row=1, col=2)
            fig.update_xaxes(title_text="Principal Axes", row=1, col=2)

            # --- ROW 1, COL 3: Variance Profiles ---
            fig.add_trace(go.Bar(x=pc_labels, y=explained_variance_ratio * 100, name="Marginal %", marker=dict(color='#ff7f0e', opacity=0.75)), row=1, col=3)
            fig.add_trace(go.Scatter(x=pc_labels, y=cumulative_variance_ratio * 100, mode='lines+markers', name='Cumulative %', line=dict(color='#d62728', width=2.5, dash='dash')), row=1, col=3)
            fig.update_yaxes(title_text="Captured Structure (%)", range=[-5, 105], row=1, col=3)
            fig.update_xaxes(title_text="Principal Axes", row=1, col=3)

            # --- ROW 2, COL 1: Remaining Noise ---
            fig.add_trace(go.Bar(x=pc_labels, y=unexplained_variance_ratio * 100, name="Excluded Info %", marker=dict(color='#2ca02c')), row=2, col=1)
            fig.update_yaxes(title_text="Excluded Info (%)", range=[-5, 105], row=2, col=1)
            fig.update_xaxes(title_text="Principal Axes", row=2, col=1)

            # --- ROW 2, COL 2: Hotelling's T^2 ---
            fig.add_trace(go.Scatter(x=k_labels, y=mean_T2_vs_k, mode='lines+markers', name='Mean T²', line=dict(color='#9467bd', width=2.5), marker=dict(size=6, symbol='diamond')), row=2, col=2)
            fig.update_yaxes(title_text="Average T² Metric", row=2, col=2)
            fig.update_xaxes(title_text="Subspace Size (k)", row=2, col=2)

            # --- ROW 2, COL 3: SPE / Q Statistic ---
            fig.add_trace(go.Scatter(x=k_labels, y=mean_Q_vs_k, mode='lines+markers', name='Mean Q (SPE)', line=dict(color='#e377c2', width=2.5), marker=dict(size=6, symbol='square')), row=2, col=3)
            fig.update_yaxes(title_text="Average Residual Energy", row=2, col=3)
            fig.update_xaxes(title_text="Truncation Cutoff (k)", row=2, col=3)

            fig.update_layout(
                title=dict(text="PCA Optimization & Feature Loading Dashboard", x=0.5, y=0.97, xanchor="center"),
                template="plotly_white",
                showlegend=True,
                legend=dict(orientation="h", yanchor="bottom", y=1.02, xanchor="center", x=0.5),
                margin=dict(t=140, b=60, l=80, r=40),
                height=750, width=1250
            )
            fig.show()

        return {
            "mean_vector": mu_hat, "covariance_matrix_S": S_matrix, "eigenvalues_lambda": lambda_hat,
            "eigenvectors_P": P_hat, "explained_variance_ratio": explained_variance_ratio,
            "cumulative_variance_ratio": cumulative_variance_ratio, "unexplained_variance_ratio": unexplained_variance_ratio,
            "transformed_scores_Z": Z_scores, "features": target_cols, "k_values": k_range,
            "mean_T2_profile": np.array(mean_T2_vs_k), "mean_Q_profile": np.array(mean_Q_vs_k)
        }

    def compute_empirical_fa(self, k: int, columns: Optional[Sequence[str]] = None, show_plot: bool = True) -> Dict[str, Any]:
        """
        Operationalizes the Factor Analysis latent subspace framework with a clean 2x2 grid.
        Properly partitions shared communality from specific uniqueness channel variances.
        """
        if self.df is None:
            raise ValueError("Error: No data loaded.")

        if columns is None:
            target_cols = self.df.select_dtypes(include=[np.number]).columns.tolist()
            if 'count' in target_cols: target_cols.remove('count')
            target_cols = list(target_cols)
        else:
            target_cols = list(columns)

        X = self.df[target_cols].copy().dropna().values
        n, m = X.shape

        # Standardize features for standard Factor Analysis (Correlation-based framework)
        mu_hat = np.mean(X, axis=0)
        std_hat = np.std(X, axis=0, ddof=1)
        Z = (X - mu_hat) / std_hat
        R_matrix = np.corrcoef(X, rowvar=False)

        # Fit Factor Analysis Model (Varimax Rotation for clean latent separation)
        fa = FactorAnalysis(n_components=k, rotation='varimax', random_state=42)
        fa.fit(Z)
        lambda_matrix = fa.components_.T
        psi_diagonal = fa.noise_variance_

        # Calculate Communality (h²) and Uniqueness (φ²) vectors
        communality = np.sum(lambda_matrix**2, axis=1)
        uniqueness = psi_diagonal

        # Map realizations using Thomson's Regression Method
        F_scores = fa.transform(Z)

        print(f"\n--- Operationalizing Latent Subspace Layer (Factor Analysis) ---")
        print(f"Decomposing structural space into {k} hidden physical factors.")
        print(f"Average System Communality (Shared Subspace Energy): {np.mean(communality)*100:.2f}%")
        print(f"Average System Uniqueness (Localized Noise Floor): {np.mean(uniqueness)*100:.2f}%")

        if show_plot:
            sensor_labels = target_cols
            factor_labels = [f"Factor {j+1}" for j in range(k)]

            fig = make_subplots(
                rows=2, cols=2, horizontal_spacing=0.20, vertical_spacing=0.28,
                subplot_titles=(
                    "Structural Loadings Matrix |λ_(j,r)|",
                    "Variance Partitioning (h² vs φ²)",
                    "Sensor Uniqueness Noise Floor Profile (φ²)",
                    "Latent Factor Scores Empirical Variance"
                )
            )

            # --- Subplot 1: Factor Loadings Heatmap ---
            fig.add_trace(
                go.Heatmap(
                    z=np.abs(lambda_matrix),
                    x=factor_labels,
                    y=sensor_labels,
                    colorscale='YlOrRd',
                    colorbar=dict(
                        title="Sensitivity",
                        x=0.42,
                        len=0.4,
                        y=0.8,
                        yanchor="middle",
                        xanchor="right"
                    ),
                    showscale=True
                ),
                row=1, col=1
            )
            fig.update_xaxes(title_text="Latent Structures", row=1, col=1)

            # --- Subplot 2: Variance Partitioning ---
            fig.add_trace(go.Bar(y=sensor_labels, x=communality * 100, name="Communality (h²)", orientation='h', marker=dict(color='#1f77b4')), row=1, col=2)
            fig.add_trace(go.Bar(y=sensor_labels, x=uniqueness * 100, name="Uniqueness (φ²)", orientation='h', marker=dict(color='#ff7f0e')), row=1, col=2)
            fig.update_layout(barmode='stack')
            fig.update_xaxes(title_text="Variance Allocation (%)", range=[0, 100], row=1, col=2)

            # --- Subplot 3: Uniqueness Line Profile ---
            fig.add_trace(go.Scatter(x=sensor_labels, y=uniqueness, mode='lines+markers', name='Uniqueness Profile', line=dict(color='#d62728', width=2, dash='dot'), marker=dict(size=8, symbol='x')), row=2, col=1)
            fig.update_yaxes(title_text="Noise Floor Scale", range=[-0.05, 1.05], row=2, col=1)
            fig.update_xaxes(title_text="Monitored Channels", row=2, col=1)

            # --- Subplot 4: Latent Factor Scores Empirical Variance ---
            factor_variances = np.var(F_scores, axis=0, ddof=1)
            fig.add_trace(go.Bar(x=factor_labels, y=factor_variances, name="Factor Variance", marker=dict(color='#2ca02c')), row=2, col=2)
            fig.update_yaxes(title_text="Variance Level", row=2, col=2)
            fig.update_xaxes(title_text="Latent Vectors", row=2, col=2)

            fig.update_layout(
                title=dict(text="Factor Analysis (FA) Latent Subspace Diagnostics Dashboard", x=0.5, y=0.97, xanchor="center"),
                template="plotly_white",
                showlegend=True,
                legend=dict(orientation="h", yanchor="bottom", y=1.02, xanchor="center", x=0.5),
                margin=dict(t=120, b=60, l=80, r=40),
                height=750, width=1100
            )
            fig.show()

        return {
            "mean_vector_mu": mu_hat, "std_vector_D": std_hat, "correlation_matrix_R": R_matrix,
            "factor_loadings_lambda": lambda_matrix, "uniqueness_psi": uniqueness,
            "communality_h2": communality, "latent_factor_scores_F": F_scores, "sensors": target_cols
        }

# =====================================================================
# DATA GENERATOR & PIPELINE EXECUTION
# =====================================================================
if __name__ == "__main__":
    np.random.seed(42)
    n_samples = 2500

    f1 = np.random.normal(0, 1, n_samples)
    f2 = np.random.normal(0, 1, n_samples)

    s1 = 0.85 * f1 + 0.10 * f2 + np.random.normal(0, 0.3, n_samples)
    s2 = 0.80 * f1 + 0.15 * f2 + np.random.normal(0, 0.35, n_samples)
    s3 = 0.12 * f1 + 0.90 * f2 + np.random.normal(0, 0.25, n_samples)
    s4 = 0.02 * f1 + 0.05 * f2 + np.random.normal(0, 1.40, n_samples)

    df_asset = pd.DataFrame(
        data=np.vstack([s1, s2, s3, s4]).T,
        columns=['Sensor_1', 'Sensor_2', 'Sensor_3', 'Sensor_4']
    )

    engine = DualSubspaceDiagnosticsEngine(df_asset)
    pca_metrics = engine.compute_empirical_pca(show_plot=True)
    fa_metrics = engine.compute_empirical_fa(k=2, show_plot=True)


--- Operationalizing Geometric De-correlation Layer (PCA) ---
Decomposing structural space of 4 features using 2500 samples.
Total System Variance (Trace[S]): 4.4674



--- Operationalizing Latent Subspace Layer (Factor Analysis) ---
Decomposing structural space into 2 hidden physical factors.
Average System Communality (Shared Subspace Energy): 50.66%
Average System Uniqueness (Localized Noise Floor): 49.35%


Question 1: The Total Variance Illusion (PCA vs. FA)
Part 1: Physical Nature of Sensor 4
A Uniqueness value ($\phi^2$) approaching $100\%$ ($>98\%$) indicates that the data generated by Sensor_4 share zero meaningful covariance with the rest of the array. Physically, this means Sensor_4 has decoupled entirely from the common physical processes ($f_1$ and $f_2$) driving the machine asset. Its signal is dominated by a localized electrical fault, telemetry saturation, or severe component degradation rather than actual plant behavior.

Part 2: The PCA Loading Contamination Mechanism
PCA operates entirely non-parametrically by maximizing global variance over the raw data space. The objective equation for the top eigenvalue is defined as:
$$\lambda_1 = \max_{\|p_1\|=1} p_1^T \Sigma p_1$$
Because Sensor_4 suffers from a massive localized white noise standard deviation ($\sigma = 1.40$), its raw variance is exceptionally large:
$$\sigma^2_{s4} \approx (1.40)^2 \approx 1.96$$
In contrast, the common physical sensors (Sensor_1 to Sensor_3) have much lower variances. To maximize global variance, the PCA engine assigns the entire weight of the primary eigenvector ($P_1$) to Sensor_4 (visible as the isolated high-intensity block in the loading matrix). This inflates $\lambda_1$ artificially, transforming an instrument error into the system's dominant "feature."

Part 3: Engineering Operational Risk
Relying strictly on an unadjusted PCA model for plant anomaly detection introduces severe risk. Because a single shorted instrument completely re-orientates the primary subspace axes, the principal control charts ($T^2$) will experience massive shifts driven entirely by sensor noise.
This tracking failure masks actual structural degradation in healthy nodes, leaving the automated pipeline completely blind to authentic system faults.

<br>
Question 2: Decoupling Structural Loading via Rotation
Part 1: Structural Simplification via Varimax Rotation
Traditional PCA enforces a strict mathematical hierarchy where eigenvectors are sequentially chosen to maximize remaining omnibus variance under rigid orthogonality constraints:
$$\lambda_1 > \lambda_2 > \dots > \lambda_m$$
This forces individual components to cross-load heavily across multiple sensors, blending separate physical factors together.

Factor Analysis relaxes this global variance requirement by applying a Varimax Rotation. Varimax performs an orthogonal rotation on the latent loading matrix ($\Lambda$) to maximize the variance of the squared loadings within each column:
$$V = \arg\max_{R} \sum_{r=1}^k \left[ \frac{1}{m} \sum_{j=1}^m \lambda_{jr}^4 - \left( \frac{1}{m} \sum_{j=1}^m \lambda_{jr}^2 \right)^2 \right]$$
This drives the weights toward either binary 1s or 0s, separating the cross-loaded signals into a clean "simple structure."

Part 2: Plant Operator Interpretability
From an operations standpoint, an unrotated PCA matrix creates a diagnostic nightmare: if an alarm triggers on PC2, the operator must decipher a complex mix of overlapping sensor inputs.
The rotated FA heatmap isolates these issues cleanly: Sensor_1 and Sensor_2 map exclusively to Factor 1 (Primary Structural Load), while Sensor_3 maps exclusively to Factor 2 (Secondary Operational Mode). This isolation allows operators to immediately pinpoint which independent process or physical asset zone is failing.
<br>
Question 3: Determining Subspace Truncation ($k$) Using $T^2$ and $Q$ Profiles
Part 1: Exact Behavior of the $Q$ Statistic Curve
As the subspace dimension moves from $k=1$ to $k=2$, the Mean $Q$ Statistic (Squared Prediction Error) plunges sharply from approximately $2.4$ down to $0.9$. This dramatic reduction indicates that the second component captures an enormous portion of the remaining un-modeled energy.
When transitioning from $k=2$ to $k=3$, the curve flattens out completely, forming a distinct "elbow" that decreases at a slow, linear rate toward $0.1$.

Part 2: Identifying the Intrinsic Dimension ($k=2$)
The combination of a sharp initial drop followed by a flat elbow identifies $k=2$ as the true hidden physical dimensionality of the system. The first dimension ($k=1$) absorbs the localized noise variance from Sensor_4. The second dimension ($k=2$) captures the authentic shared structural variance binding Sensor_1, Sensor_2, and Sensor_3.
Beyond $k=2$, the remaining components offer negligible variance reduction, showing that all shared structural information has been fully exhausted.

Part 3: The Danger of Over-fitting to $k=3$
Setting the truncation threshold to $k=3$ forces the residual noise space into the "clean" reference subspace model. Because this additional dimension tracks minor, localized sensor errors rather than real structural phenomena, the model overfits the asset's baseline.
Consequently, Hotelling's $T^2$ control limits will continuously trigger false alarms in response to normal, everyday sensor noise.
<br>
Question 4: Operational Trade-offs in System Health Monitoring
Structural Comparison Matrix
To help evaluate deployment options for an industrial predictive maintenance pipeline, the baseline trade-offs

| Operational Dimension | PCA Pipeline (T2 & Q Control Charts) | FA Pipeline (Rotated Latent Score Tracking) |
| :-------------------- | :----------------------------------- | :------------------------------------------ |
| Fault Isolation Speed   | Low (requires manual post-alarm analysis of the $Q$ contribution vector). | Instantaneous (the specific factor score chart deviates directly). |
| Subspace Stability    | Vulnerable (extreme single-sensor noise alters the base tracking axes). | Highly Robust (orthogonal rotation isolates anomalies to specific nodes). |
| Computational Footprint | Minimal (simple linear matrix operations; fast execution). | Moderate (requires expectation-maximization or iterative regression fits). |
| Noise Resilience      | Poor (mixes unique channel noise into the main structural model). | Excellent (absorbs localized channel errors into uniqueness $\Psi$). |